### ISSUE TITLE
Charset need to be passed as a parameter also in unpack methods

### ISSUE THREAD

**Issue by fbmanhani** [zeroturnaround/zt-zip#79]
I've found a bug while using the method unpackEntry(InputStream is, String name, File file). When no charset is set, it throws an exception when unpacking zips with special named files.

**Comment by toomasr** [comment]
Do you have an archive and sample code to reproduce this issue. A quick test I tried didn't result in an exception.

Also, what is the expected behaviour? Fallback to some default charset?

**Comment by fbmanhani** [comment]
Basically, i did something like this:

```
ZipUtil.iterate(new ByteArrayInputStream(content), new ZipEntryCallback() {
					@Override
					public void process(InputStream in, ZipEntry zipEntry) throws IOException {
						if(!zipEntry.isDirectory()){
							String name = getName(code, zipEntry);
							File file = new File(name);
							**ZipUtil.unpackEntry(new ByteArrayInputStream(content), zipEntry.getName(), file);**
							if(!hasPRJ){
								hasPRJ = zipEntry.getName().endsWith(".prj");
							}
						}
					}
				}, CP866);

```
The point where i get the exception is at the unpack part of the code. Note that CP866 is passed as a parameter to the iterate method but unpackEntry doesn't have a charset parameter.

The sample file is attached. As you will see, it has some special characters at the file names.
[Biomas_AMZ.zip](https://github.com/zeroturnaround/zt-zip/files/339351/Biomas_AMZ.zip)

**Comment by cordje** [comment]
I believe I have replicated the issue, (although the problem description was not crystal clear).

This test case fails on the second assertTrue. A solution is shown by the code preceding the first assertTrue.

```
  public void testIterateExtractWithCP866Filenames() throws URISyntaxException, IOException {

    final File src = new File("C:\\Users\\R\\workspace\\GOAL\\zt-zip\\src\\test\\resources\\Biomas_AMZ.zip");
    FileInputStream inputStream = new FileInputStream(src);

    ZipUtil.iterate(inputStream, new ZipEntryCallback() {
      public void process(InputStream in, ZipEntry zipEntry) throws IOException {
        File tmp1 = File.createTempFile(zipEntry.getName(), "bleh");
        ZipUtil.unpackEntry(new ZipFile(src, Charset.forName("CP866")), zipEntry.getName(), tmp1);
        assertTrue(tmp1.exists() && tmp1.length() > 0);

        File tmp2 = File.createTempFile(zipEntry.getName(), "bleh");
        ZipUtil.unpackEntry(src, zipEntry.getName(), tmp2);
        assertTrue(tmp2.exists() && tmp2.length() > 0);
      }
    }, Charset.forName("CP866"));

    inputStream.close();
  }
```

The problem is that ZipUtils unpackEntry() methods do not receive a Charset parameter, and without it Java cannot unzip the files in Biomas_AMZ.zip.
More specifically, in the code provided by @fbmanhani, the problematic unpackEntry() call has the following call trace:

```
unpackEntry(InputStream is, String name, File file)
handle(is, name, new FileUnpacker(file))
iterate(is, new SingleZipEntryCallback(name, action))
```

To fix the issue, the iterate() method at the bottom should instead call the (already implemented): 
`handle(InputStream is, String name, ZipEntryCallback action, Charset charset)`

This requires adding the charset parameter to the unpackEntry() and handle() methods in ZipUtils.

**Comment by OneWorld123** [comment]
I stumpled over the issue as well. I unpack archives created on Windows on Android devices. The filenames of the zipped files contain special characters (Ä, Ö, Ü in my case). Unfortunately äöüß.txt turns into ����.txt after unpacking using `ZipUtil.unpack(myArchive, unpackDir);`   :-(

Please let me pass the desired character encoding in `ZipUtil.unpack(myArchive, unpackDir);`

**Comment by toomasr** [comment]
Hmm, it seems the case has fallen through the cracks.

@cordje thanks for the great example. I think I do understand it. In the first example you use a ZipFile which takes a Charset and you can call the unpack entry because getName() will match. If there is no charset it won't work.

Why wouldn't you just use the `InputStream in` to write out the file to tmp2? No need to use the `ZipUtil.unpackEntry`. From the API point of view I agree that it should propagate the Charset to the `process` method.

@OneWorld123 I'll try to reproduce your issue as well.

**Comment by toomasr** [comment]
I've added the necessary methods. @OneWorld123 @cordje @fbmanhani do you mind taking a look at the changes before I prep a release?

**Comment by fbmanhani** [comment]
@toomasr couldn't check the code in execution now, but, the changes you've made will solve the problem i've mentioned.  Thanks for the effort and sorry if i didn't make the problem clear at first.

**Comment by cordje** [comment]
Looks good to me, thanks.

**Comment by OneWorld123** [comment]
Unfortunately `ZipFile(File file, int mode, Charset charset) ` is a constructer added in JDK1.7. Android added JDK1.7 support just recently to Android 7.0 (Nougat). However, I have to support devices from Androd 4.2.2 onwards ;(

- Android doc: https://developer.android.com/reference/java/util/zip/ZipFile.html#pubctors
- Work around http://stackoverflow.com/questions/4085293/zipentry-file-name-encoding
- Other users on SO claim that this lib just works http://stackoverflow.com/questions/24509264/extracting-file-with-zipinputstream-error-utfdataformatexception/24510190#24510190

It would be best if the zip file had details about its charset in its header. But according to https://marcosc.com/2008/12/zip-files-and-encoding-i-hate-you/ that's not the case. The comments on that patch also talk about an auto detect feature of the actual charset used.

**Comment by toomasr** [comment]
@OneWorld123 so maybe your solution is to iterate as you would and then writing out the bytes from the Inputstream to a filename of your choosing? To get the proper filesnames though will be tougher, probably checking the internals of the ZipEntry/ZipFile implementations and through reflections maybe can get access to the bytes of the filename and then decode with proper encoding.

**Comment by toomasr** [comment]
I released 1.11 version of zt-zip yesterday. It brings the Charset to the unpack methods. I'm closing this issue.
